pkg/dns-server/resolver/server_test.go

```diff
@@ -0,0 +1,202 @@
+package resolver
+
+import (
+	"fmt"
+
+	"github.com/miekg/dns"
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+var _ = Describe("DNS server", func() {
+
+	It("DNS Server basic functionality", func(done Done) {
+		resolver, err := NewSimpleDNSResolver("127.0.0.1", "5653", "240.0.0.0/4")
+		Expect(err).ToNot(HaveOccurred())
+
+		err = resolver.AddDomain(".kuma")
+		Expect(err).ToNot(HaveOccurred())
+
+		_, err = resolver.AddServiceToDomain("service", "kuma")
+		Expect(err).ToNot(HaveOccurred())
+
+		ip, err := resolver.ForwardLookup("service.kuma")
+		Expect(err).ToNot(HaveOccurred())
+
+		service, err := resolver.ReverseLookup(ip)
+		Expect(err).ToNot(HaveOccurred())
+		Expect(service).To(Equal("service.kuma"))
+
+		// ready
+		close(done)
+	})
+
+	It("DNS Server network functionality", func(done Done) {
+		resolver, err := NewSimpleDNSResolver("127.0.0.1", "5653", "240.0.0.0/4")
+		Expect(err).ToNot(HaveOccurred())
+
+		err = resolver.AddDomain(".kuma")
+		Expect(err).ToNot(HaveOccurred())
+
+		_, err = resolver.AddServiceToDomain("service", "kuma")
+		Expect(err).ToNot(HaveOccurred())
+
+		ip, err := resolver.ForwardLookup("service.kuma")
+		Expect(err).ToNot(HaveOccurred())
+
+		stop := make(chan struct{})
+		go func() {
+			err := resolver.Start(stop)
+			Expect(err).ToNot(HaveOccurred())
+		}()
+
+		client := new(dns.Client)
+		message := new(dns.Msg)
+		_ = message.SetQuestion("service.kuma.", dns.TypeA)
+		var response *dns.Msg
+		Eventually(func() error {
+			response, _, err = client.Exchange(message, "127.0.0.1:5653")
+			return err
+		}).ShouldNot(HaveOccurred())
+		Expect(response.Answer[0].String()).To(Equal(fmt.Sprintf("service.kuma.\t3600\tIN\tA\t%s", ip)))
+
+		message = new(dns.Msg)
+		_ = message.SetQuestion("backend.kuma.", dns.TypeA)
+		response, _, err = client.Exchange(message, "127.0.0.1:5653")
+		Expect(err).ToNot(HaveOccurred())
+		Expect(len(response.Answer)).To(Equal(0))
+
+		// stop the resolver
+		stop <- struct{}{}
+
+		// ready
+		close(done)
+	})
```

Comment:
You test here two different use cases.
1) DNS server responds with IP for service it has
2) DNS server responds with nothing for service it does not know.

I'd go with starting the server in BeforeEach and wait for it to start and then build use cases tests on top of it.